compiler/x/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3216,6 +3215,252 @@ TR::Register *OMR::X86::TreeEvaluator::arraycopyEvaluator(TR::Node *node, TR::Co
+      generateRegMemInstruction(LEARegMem(), node, srcReg, generateX86MemoryReference(srcReg, sizeReg, 0, -(intptr_t)elementSize, cg), cg);
+      generateRegMemInstruction(LEARegMem(), node, dstReg, generateX86MemoryReference(dstReg, sizeReg, 0, -(intptr_t)elementSize, cg), cg);
+      generateInstruction(STD, node, cg);
+      generateRepMovsInstruction(REPMOVS, node, sizeReg, NULL, cg);
+      generateInstruction(CLD, node, cg);
+      generateLabelInstruction(JMP4, node, MainEndLabel, cg);
+      BackwardPath->swapInstructionListsWithCompilation();
+
+      generateLabelInstruction(LABEL, node, MainEndLabel, dependencies, cg);
+      }
+   }
+
+TR::Register *OMR::X86::TreeEvaluator::arraycopyEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   static bool UseDeprecatedArraycopy = feGetEnv("TR_UseDeprecatedArraycopy");
+   if (UseDeprecatedArraycopy)
+      {
+      return deprecated_arraycopyEvaluator(node, cg);
+      }
+   if (node->isReferenceArrayCopy() && !node->isNoArrayStoreCheckArrayCopy())
+      {
+      return TR::TreeEvaluator::VMarrayStoreCheckArrayCopyEvaluator(node, cg);
+      }
+   // There are two cases.
+   // In the first case we know that a simple memmove or memcpy operation can
+   // be done. For this case there are 3 children:
+   //    1) The byte source pointer
+   //    2) The byte destination pointer
+   //    3) The byte count
+   //
+   // In the second case we must generate run-time tests to see if a simple
+   // byte copy can be done or if an element-by-element copy is needed.
+   // For this case there are 5 children:
+   //    1) The original source object reference
+   //    2) The original destination object reference
+   //    3) The byte source pointer
+   //    4) The byte destination pointer
+   //    5) The byte count
+   //
+
+   // ALL nodes need be evaluated to comply argument evaluation order;
+   // since size node is the last node, its evaluation can be delayed for further optimization
+   TR::Node* sizeNode = node->getLastChild();
+   for (int i = 0; i < node->getNumChildren()-1; i++)
+      {
+      cg->evaluate(node->getChild(i));
+      }
+
+   TR::Register* srcReg = cg->allocateRegister();
+   TR::Register* dstReg = cg->allocateRegister();
+   generateRegRegInstruction(MOVRegReg(), node, srcReg, node->getChild(node->getNumChildren() - 3)->getRegister(), cg);
+   generateRegRegInstruction(MOVRegReg(), node, dstReg, node->getChild(node->getNumChildren() - 2)->getRegister(), cg);
+
+   TR::DataType dt = node->getArrayCopyElementType();
+   uint32_t elementSize = 1;
+   static bool ForceByteArrayElementCopy = feGetEnv("TR_ForceByteArrayElementCopy");
+   if (!ForceByteArrayElementCopy)
+      {
+      if (node->isReferenceArrayCopy() || dt == TR::Address)
+         elementSize = TR::Compiler->om.sizeofReferenceField();
+      else
+         elementSize = TR::Symbol::convertTypeToSize(dt);
+      }
+
+   bool isShortConstantArray = false;
+   if (isShortConstantArray)
```

Comment:
Thank you for point this out. It is a placeholder for a subsequent PR, which will further optimize cases that array length is known at compilation.